p2p/src/peers/mod.rs

```diff
@@ -4,43 +4,113 @@ use rand::{thread_rng, Rng};
 use serde::{Deserialize, Serialize};
 use std::{
     cmp,
-    collections::HashMap,
+    collections::{HashMap, HashSet},
+    convert::TryFrom,
     fmt,
+    hash::{Hash, Hasher},
     net::{IpAddr, Ipv4Addr, SocketAddr},
+    time::Duration,
 };
 
 use rand::seq::IteratorRandom;
 use witnet_crypto::hash::calculate_sha256;
 use witnet_util::timestamp::get_timestamp;
 
 /// Peer information being used while listing available Witnet peers
-#[derive(Serialize, Deserialize)]
-struct PeerInfo {
-    address: SocketAddr,
-    timestamp: i64,
+#[derive(Debug, Deserialize, Eq, Serialize)]
+pub struct PeerInfo {
+    /// The socket address for a potential peer
+    pub address: SocketAddr,
+    /// Last time that the peer address was tried
+    pub timestamp: i64,
+}
+
+/// This fuzzy implementation of `PartialEq` does the magic of comparing `PeerInfo` by only the
+/// address and not any other field.
+///
+/// # Examples
+/// ```rust
+/// use witnet_p2p::peers::PeerInfo;
+/// use std::{hash::Hash, net::SocketAddr, str::FromStr};
+///
+/// let peer_1 = PeerInfo {
+///     address: SocketAddr::from_str("127.0.0.1:21337").unwrap(),
+///     timestamp: 0,
+/// };
+/// let peer_2 = PeerInfo {
+///     address: SocketAddr::from_str("127.0.0.1:21337").unwrap(),
+///     timestamp: 1,
+/// };
+///
+/// assert_eq!(peer_1, peer_2);
+/// ```
+impl PartialEq for PeerInfo {
+    fn eq(&self, other: &Self) -> bool {
+        self.address == other.address
+    }
+}
+
+/// Forced implementation of `Hash` for `PeerInfo` so that the famous hash comparison property
+/// does actually hold:
+///
+/// `k1 == k2 ⇒ hash(k1) == hash(k2)`
+///
+/// # Examples
+/// ```rust
+/// use witnet_p2p::peers::PeerInfo;
+/// use std::{collections::hash_map::DefaultHasher, hash::{Hash, Hasher}, net::SocketAddr, str::FromStr};
+///
+/// let mut hash_1 = DefaultHasher::new();
+/// PeerInfo {
+///     address: SocketAddr::from_str("127.0.0.1:21337").unwrap(),
+///     timestamp: 0,
+/// }.hash(&mut hash_1);
+/// let mut hash_2 = DefaultHasher::new();
+/// PeerInfo {
+///     address: SocketAddr::from_str("127.0.0.1:21337").unwrap(),
+///     timestamp: 1,
+/// }.hash(&mut hash_2);
+/// assert_eq!(hash_1.finish(), hash_2.finish());
+/// ```
+impl Hash for PeerInfo {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.address.hash(state)
+    }
+}
+
+impl From<&SocketAddr> for PeerInfo {
+    fn from(addr: &SocketAddr) -> Self {
+        PeerInfo {
+            address: *addr,
+            timestamp: Default::default(),
+        }
+    }
 }
 
 /// Peers TBD
-#[derive(Default, Serialize, Deserialize)]
+#[derive(Default, Deserialize, Serialize)]
 pub struct Peers {
-    /// Bucket for tried addresses
-    tried_bucket: HashMap<u16, PeerInfo>,
+    /// Bucket for "iced" addresses (will not be tried in a while)
+    ice_bucket: HashSet<PeerInfo>,
```

Comment:
This looks like it should be a `HashMap<SocketAddr, u64>`, where `u64` is the timestamp. I would prefer to do it that way to so we can remove the custom implementation of `PartialEq`, because that's dangerous and it may cause us problems in the future. Not sure about the performance differences between `HashSet<PeerInfo>` + custom `PartialEq`, and `HashMap<SocketAddr, u64>`.

Reply:
True, that's not needed anymore :+1: